include/ghex/transport_layer/ucx/communicator.hpp

```diff
@@ -113,11 +112,9 @@ namespace gridtools {
                     template <typename Message>
                     [[nodiscard]] future<void> recv(Message &msg, rank_type src, tag_type tag)
                     {
-                        const auto rtag = ((std::uint_fast64_t)tag << 32) | 
+                        const auto rtag = ((std::uint_fast64_t)tag << 32) |
                                            (std::uint_fast64_t)(src);
-                        return m_send_worker->m_thread_primitives->critical(
-                            [this,rtag,&msg,src,tag]()
-                            {
+                        std::lock_guard<decltype(m_send_worker->mutex())> lock(m_send_worker->mutex());
```

Comment:
this will only work for the spin-lock recursive mutex. A standard mutex is not recursive and we'll need a `unique_lock` for that. If we stick to our custom mutex for ucx, then everything is fine.

Reply:
I should just put the type instead of `decltype`, so if we change something we trigger an error and we check

Reply:
yes, sounds good

Reply:
shouldn't we use the mutex from the recv worker? since the send worker is not shared?
also shouldn't there be a `std::remove_reference_t` in the `decltype`?